CARDS-1783: Make email scheduling more flexible

IntegralProgrammer opened the pull request:
The PR removes the `InitialNotificationsTask` and `ReminderNotificationsTask` scheduled tasks and provides, as a replacement, an OSGi component that can be configured and instantiated to provide emails for upcoming appointments. The functionality of the `InitialNotificationsTask` and `ReminderNotificationsTask` scheduled tasks is preserved by the `InitialNotificationsTask` and `ReminderNotificationsTask` configurations for `io.uhndata.cards.proms.emailnotifications.UpcomingNotificationsFactory` that is loaded upon repository initialization.

To test:

- Build this PR with `mvn clean install`
- Follow the instructions in https://github.com/data-team-uhn/cards/blob/dev/Utilities/Development/EmailServer/README.md to setup a testing mail server and client to see the emails sent by CARDS. Follow the _Using (with `./start_cards.sh`)_ instructions as opposed to the _Using (with Docker Compose)_ instructions.
- Open a new terminal in `Utilities/Development` and start the mock Torch server by running `nodejs mock_torch.js --appointment-time-hours-from-now=24`.
- Open a new terminal in `Utilities/Development` and start the mock Slack server by running `nodejs mock_slack.js`
- Start CARDS in _Cards4Proms_ mode with the command `PATIENT_NOTIFICATION_FROM_ADDRESS=datapro@uhn.ca PATIENT_NOTIFICATION_FROM_NAME=DATAPRO NIGHTLY_NOTIFICATIONS_SCHEDULE="0 * * * * ? *" CARDS_HOST_AND_PORT='localhost:8080' NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE="0 * * * * ? *" SLACK_PERFORMANCE_URL="http://localhost:8000/" ./start_cards.sh --dev --project cards4proms -V emailnotifications.smtps.checkserveridentity=false`.
- Visit `http://localhost:8080` and login as `admin`:`admin`.
- Visit `http://localhost:8080/system/console/configMgr` and add a new _PROMs Import_ configuration with the following values:
  - `Name`: `LocalNodeJS`
  - `Import schedule`: `0 0 0 * * ? *`
  - `days to query`: `3`
  - `Endpoint URL`: `http://localhost:8011/`
  - `Authentication URL`: (leave blank)
  - `Vault token`: (leave blank)
  - `Clinic names`: `6012-HC-Congenital Cardiac`
  - `Provider names`: `SomeParticipantEID`
  - `Allowed provider roles`: `ATND`
  - `Vault role name`: (leave blank)
  - `Dates to query`: (leave blank)
- Trigger the import by visiting `http://localhost:8080/Subjects.importTorch?config=LocalNodeJS`
- Verify that the `12345 / AppointmentOneFhirID-Cardio` appointment was created.
- You should now receive an email with the subject _"Reminder: You have 24 hours left to complete your pre-appointment questions"_ once every minute.
- Visit `http://localhost:8080/system/console/configMgr` once again and create a new _Upcoming appointment notification_ with the following values:
  - `Name`: `TestNotifications`
  - `Metric Name`: `Test Notification emails sent`
  - `Clinic Mapping Path`: `/Proms/ClinicMapping/853703519`
  - `Email Subject Line`: `Email Notification System Test`
  - `Plaintext Email Template JCR Path`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.txt`
  - `HTML Email Template JCR Path`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.html`
  - `Days to the visit`: `1`
- Every minute you should receive an upcoming appointment notification email with the subject line `Email Notification System Test`.
- On the mock Slack server, you should also be receiving, once every minute, performance notifications that include an incrementing count of `Test Notification emails sent` and `Reminder Emails Sent`. It should also display `1` for the `Total` number of `Imported Appointments`.
- Repeat the above steps but, this time, specify an appointment 72 hours from now instead of 24 hours when starting the mock Torch server (ie. `nodejs mock_torch.js --appointment-time-hours-from-now=72`). Verify that you are receiving an email with the subject line _"Welcome to DATAPRO: Answer your Pre-Appointment Questions"_ once every minute. Also verify that the `Initial Emails Sent` count in the Slack notifications is auto-incrementing.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Create the GeneralUpcomingNotification class and associated OSGi configuration factory

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Add logging to the GeneralUpcomingNotification class

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Now actually schedules the upcoming appointment notification tasks

TODO: Need to be able to unschedule the tasks

Renamed the GeneralUpcomingNotification class to UpcomingNotificationsFactory

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Implemented unscheduling of upcoming appointment notification tasks

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Keep performance metrics of each scheduled notification job

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Allow for the cards-metrics service user to create child nodes under /Metrics/ as this is required whenever a new email notifications configuration is instantiated - we would like to keep track of its performance statistics

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Introduce the clinicId parameter when sending notifications, which, if it is not null, can be used to find appointments for only a specific clinic identifier as opposed to all clinics.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Remove the InitialNotificationsTask and the ReminderNotificationsTask. They can be implemented using OSGi configuration.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename "ClinicID" OSGi configuration parameter to "Clinic Mapping Path" and update its description.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Remove auto-creation of /Metrics/InitialEmailsSent and /Metrics/ReminderEmailsSent

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Create the "name" JCR nodes for the TotalSurveysSubmitted, AppointmentServeysSubmitted, and ImportedAppointments metrics upon repository initialization

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Create the "name" JCR node under /Metrics/ when adding a new email notifications task via OSGi

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Create the Metrics.getHumanName method

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Replace string "value" with string constant PROP_VALUE in Metrics.java

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Generalize SlackNotificationsTask so that it reports all metrics recorded in the JCR under /Metrics/ as opposed to only the five original hard-coded performance metrics.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Clean up code and add Javadoc for the Metrics.createStatistic method

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Add the Initial Notification (72 hour) and Reminder Notification (24 hour) tasks to the proms-resources feature.json file so that they are automatically configured upon repository initialization

acrowthe commented:
Code reviewed, not tested

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/AbstractPromsNotification.java:
```diff
@@ -94,16 +101,28 @@ public long sendNotification(final int daysInTheFuture, final String emailTextTe
+                 */
+                String emailTextTemplate = null;
+                String emailHtmlTemplate = null;
+                if (clinicId != null) {
+                    emailTextTemplate = AppointmentUtils.getEmailTemplateFromPath(resolver, emailTextTemplateName);
+                    emailHtmlTemplate = AppointmentUtils.getEmailTemplateFromPath(resolver, emailHtmlTemplateName);
```
I was a bit surprised that `AppointmentUtils#getClinicEmailTemplate` is called (evetually) when `clinic == null`, and when `clinic != null` we call `getEmailTemplateFromPath` which completely ignores the clinic.

I think that the `clinic == null` case is just legacy code that isn't actually active at runtime, should it just be deleted?

IntegralProgrammer replied:
Yes, the `clinic == null` case is, essentially, legacy code that would no longer typically be used. I actually just tried right now to specify an _Upcoming appointment notification_ with a missing `clinicId`. The `else` branch is taken as `clinicId == null` but a NullPointerException occurs as `AppointmentUtils.getVisitEmailTemplate` calls `AppointmentUtils.getVisitClinic` which calls `AppointmentUtils.getQuestionAnswerForSubject` and this results in a NPE at

https://github.com/data-team-uhn/cards/blob/e3a1d0e9b7d7ab5ae5c193133feadee3fad9eadd/proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/AppointmentUtils.java#L114

because

https://github.com/data-team-uhn/cards/blob/e3a1d0e9b7d7ab5ae5c193133feadee3fad9eadd/proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/AppointmentUtils.java#L215-L228

attempts to access the `/Questionnaires/Visit information/surveys` JCR node which no longer exists - I believe that this has been replaced by the `/Questionnaires/Visit information/clinic` node.

Therefore, should this code be updated to handle the `/Questionnaires/Visit information/surveys` --> `/Questionnaires/Visit information/clinic` change or should the handling of the `clinic == null` case be removed altogether?

@sdumitriu @marta-

marta- commented:
@IntegralProgrammer @acrowthe does this generalization allow sending notification emails AFTER the visit time? For prems, the notifications to fill out experience surveys must be sent a couple of days after the patient is discharged.

marta- reviewed (dismissed):
Looks like sending notification emails to past visits might actually work if `daysBeforeVisit` in `GeneralNotificationsTask` is negative (via @IntegralProgrammer; to be confirmed by testing), which means the PREMs use case might be covered functionality-wise.

That being said, since we want to generalize this code for both proms and prems, some more minor refactoring might make the code easier to digest:
- [ ] Rename `AbstractPromsNotification` to `AbstractEmailNotification` (or `AbstractPatientNotification`?)
- [ ] `In Abstract****Notification`, rename `daysInTheFuture` to `differenceInDays` and document that a negative value sends emails for past events, while a positive value sends emails for future events
- [ ] Similarly in GeneralNotificationTask rename `daysBeforeVisit` to `daysToVisit` and document the meaning of positive and negative values.
- [ ] In `UpcomingNotificationFactory`, rename `daysBeforUpcomingVisit` to `daysToVisit`, change the name and description and document the meaning of positive and negative values.
Can we assume that in the name of this class `Upcoming` refers to the notification about to be sent, not to the visit in the future? if not, this class also needs to be renamed.

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the AbstractPromsNotification class to AbstractEmailNotification

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the daysInTheFuture parameter to differenceInDays and add javadoc

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the daysBeforeVisit parameter to daysToVisit and add javadoc

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the daysBeforeUpcomingVisit OSGi configuration parameter to daysToVisit and update the description of it

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the UpcomingNotificationsFactory class to AppointmentEmailNotificationsFactory

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the UpcomingNotificationsFactory class to AppointmentEmailNotificationsFactory

IntegralProgrammer committed:
CARDS-1783: Make email scheduling more flexible

Rename the daysBeforeUpcomingVisit to daysToVisit

IntegralProgrammer commented:
> Looks like sending notification emails to past visits might actually work if `daysBeforeVisit` in `GeneralNotificationsTask` is negative (via @IntegralProgrammer; to be confirmed by testing), which means the PREMs use case might be covered functionality-wise.
> 
> That being said, since we want to generalize this code for both proms and prems, some more minor refactoring might make the code easier to digest:
> 
>     * [ ]  Rename `AbstractPromsNotification` to `AbstractEmailNotification` (or `AbstractPatientNotification`?)
> 
>     * [ ]  `In Abstract****Notification`, rename `daysInTheFuture` to `differenceInDays` and document that a negative value sends emails for past events, while a positive value sends emails for future events
> 
>     * [ ]  Similarly in GeneralNotificationTask rename `daysBeforeVisit` to `daysToVisit` and document the meaning of positive and negative values.
> 
>     * [ ]  In `UpcomingNotificationFactory`, rename `daysBeforUpcomingVisit` to `daysToVisit`, change the name and description and document the meaning of positive and negative values.
>       Can we assume that in the name of this class `Upcoming` refers to the notification about to be sent, not to the visit in the future? if not, this class also needs to be renamed.

I have done these refactorings and can confirm that tasks can be created that send emails both for appointments in the future (PROMS) as well as appointments in the past (PREMS).

To test:

- Follow the original testing instructions in the description of this PR and ensure that the default _72 hour initial notification_ and _24 hour reminder notification_ tasks still work as expected.
- Visit http://localhost:8080/system/console/configMgr and create a new _Appointment email notification_ with the following parameters:
  - `name`: `PastAppointmentFollowUp`
  - `metricName`: `Follow-up Emails Sent`
  - `clinicId`: `/Proms/ClinicMapping/853703519`
  - `emailSubject`: `How was your appointment?`
  - `plainTextEmailTemplatePath`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.txt`
  - `htmlEmailTemplatePath`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.html`
  - `daysToVisit`: `-7`
- Visit http://localhost:8080/bin/browser.html and edit a _Visit information_ Form so that the visit date is 1 week ago from today (you'll need to check-out the Form node before doing this and check it back in after)
- Ensure that you are receiving emails with the subject `How was your appointment?` once every minute.

acrowthe reviewed (approved):
Reviewed and tested
